GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.cs

```diff
@@ -0,0 +1,105 @@
+        public StorageClient(StorageService service)
+        {
+            Service = service;
+        }
+
+        /// <summary>
+        /// Constructs a new client by creating a <see cref="StorageService"/> which uses the
+        /// given <see cref="BaseClientService.Initializer"/> for request initialization.
+        /// </summary>
+        /// <param name="initializer">The initializer to use in the service. Must not be null.</param>
+        public StorageClient(BaseClientService.Initializer initializer) : this(new StorageService(initializer))
+        {
+        }
+
+        /// <summary>
+        /// Asynchronously creates a new <see cref="StorageClient"/> from the default application credentials.
+        /// </summary>
+        /// <remarks>
+        /// The application credentials are available to developers by running <c>gcloud auth</c> from the
+        /// command line, and are available automatically on Google Cloud Platform hosts.
+        /// </remarks>
+        /// <param name="applicationName">The name of the application to create a client for. Must not be null.</param>
+        /// <returns>A client configured with the application-default credentials.</returns>
+        public static async Task<StorageClient> FromApplicationCredentials(string applicationName)
+        {
+            var initializer = new BaseClientService.Initializer
+            {
+                HttpClientInitializer = await GoogleCredential.GetApplicationDefaultAsync(),
+                ApplicationName = applicationName,
+            };
+            return new StorageClient(initializer);
+        }
+
+        /// <summary>
+        /// Asynchronously lists all the buckets for a given project.
+        /// </summary>
+        /// <remarks>
+        /// This is a convenience method for calling <see cref="ListAllBucketsAsync(string, CancellationToken)"/>.
+        /// </remarks>
+        /// <param name="project">The project to list the buckets from. Must not be null.</param>
+        public Task<List<Bucket>> ListAllBucketsAsync(string project)
+        {
+            return ListAllBucketsAsync(project, cancellationToken: default(CancellationToken));
+        }
+
+        /// <summary>
+        /// Asynchronously lists all the buckets for a given project.
+        /// </summary>
+        /// <remarks>
+        /// This lists all the buckets within a project before the returned task completes.
+        /// This does not support reporting progress, or streaming the results.
+        /// </remarks>
+        /// <param name="project">The project to list the buckets from. Must not be null.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A list of all buckets within the project.</returns>
+        public async Task<List<Bucket>> ListAllBucketsAsync(string project, CancellationToken cancellationToken)
```

Comment:
Could do... but I'm not sure how much benefit there'd be. If we always *do* return a `List<T>`, and can't think of any reason why we'd use any other implementation, we don't really need to hide that from consumers - who *may* want to use list-specific members.

I'm definitely up for other naming suggestions, btw - bearing in mind that other list operations would be:

- The original "perform one RPC to get a page"
- List all buckets asynchronously, but lazily

(Both of those may have synchronous versions too - we'll definitely have the lazy form, but we may not need this list-returning form as the caller can always call `ToList()` on an `IEnumerable<T>`.)